xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -0,0 +1,300 @@
+      xdsRequestWriter.onNext(edsRequest);
+    }
+
+    // Must run in SynchronizationContext
+    void shutdown() {
+      if (cancelled) {
+        return;
+      }
+      cancelled = true;
+      xdsRequestWriter.onError(
+          Status.CANCELLED.withDescription("xDS Client shutdown").asRuntimeException());
+    }
+
+    final class XdsResponseReader implements StreamObserver<DiscoveryResponse> {
+      // Must be accessed in SynchronizationContext
+      boolean firstEdsResponseReceived;
+
+      @Override
+      public void onNext(final DiscoveryResponse value) {
+
+        class HandleResponseRunnable implements Runnable {
+
+          @Override
+          public void run() {
+            String typeUrl = value.getTypeUrl();
+            if (EDS_TYPE_URL.equals(typeUrl)) {
+              ClusterLoadAssignment clusterLoadAssignment;
+              try {
+                // maybe better to run this deserialization task out of syncContext?
+                clusterLoadAssignment =
+                    value.getResources(0).unpack(ClusterLoadAssignment.class);
+              } catch (InvalidProtocolBufferException | RuntimeException e) {
+                if (logger.isLoggable(Level.INFO)) {
+                  logger.log(
+                      Level.INFO, "Unable to parse clusterLoadAssignment from EDS response" + value,
+                      e);
+                }
+                // TODO(zdapeng): NACK
+                return;
+              }
+
+              logger.log(
+                  Level.FINE,
+                  "Received an EDS response: {0}",
+                  clusterLoadAssignment);
+
+              String cluster = clusterLoadAssignment.getClusterName();
+
+              // TODO(zdapeng): Remove this when we can obtain the right cluster name.
+              if (endpointWatchers.containsKey(authority)) {
+                cluster = authority;
+              }
+
+              if (endpointWatchers.containsKey(cluster)) {
+                firstEdsResponseReceived = true;
+                endpointWatchers.get(cluster).onEndpointUpdate(clusterLoadAssignment);
+              }
+            }
+          }
+        }
+
+        syncCtx.execute(new HandleResponseRunnable());
+      }
+
+      @Override
+      public void onError(final Throwable t) {
+        logger.log(Level.INFO, "XdsClient received an error", t);
+
+        syncCtx.execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                closed = true;
+                if (cancelled) {
+                  return;
+                }
+                for (EndpointWatcher endpointWatcher : endpointWatchers.values()) {
+                  endpointWatcher.onError(t);
+                }
+                scheduleRetry();
+              }
+            });
+      }
+
+      @Override
+      public void onCompleted() {
+        onError(Status.INTERNAL.withDescription("Server closed the ADS streaming RPC")
```

Comment:
This doesn't seem correct. It gives a false illusion of receiving two responses.